### ISSUE TITLE
Return the attention weights using the Transformer Encoder class. 

### ISSUE THREAD

**Issue by PaulOppelt** [pytorch/pytorch#88810]
### 🚀 The feature, motivation and pitch

Hey, when using the Transformer Encoder class it would be nice to have ```need_weights``` as an additional variable that can be passed down to ```F.multi_head_attention_forward ```. Right now the feature of returning the attention weights is as far as my understanding goes not accessible from the Transformer class since ```need_weights``` is manually set to false. It would be great if this was possible since one could very easily compute the attention rollout.

**Comment by erichan1** [comment]
Hi Paul! Thanks for bringing up the issue. I think we have attn_weights for TransformerEncoder planned at some point, but not sure on the eta...

**Comment by mettij** [comment]
Hi, im interested in picking up this issue for a class along with a partner. Is there anyway we can get more on information on this @erichan1 ?

**Comment by erichan1** [comment]
@mettij This is a reasonably substantial change. I'd estimate at least a couple of weeks for back and forth? And since it's an API change there will be some level of scrutiny from the PyTorch team you'll have to wait on. If you're ok with that, then yes a PR would be nice.

Relevant bit of code is here:
https://github.com/pytorch/pytorch/blob/a6d72f44a4e8b6e9d2e878f30fd8b1d3e1197f0e/torch/nn/modules/transformer.py#L541.

Rough steps:
1. Add a need_weights arg to TransformerEncoder
2. Pass it into sa_block, return your attn_weights along with regular output only if needs_weights == True.
3. Make sure to disable fast path when need_weights==True by adding a condition in why_not_sparsity_fast_path. PyTorch transformer has a fast inference path. You could actually make need_weights work for the fast path by hooking up a couple of things, but let's skip it for now. Just default to slow path when we have this arg. 
4. Add a test (and probably expand existing tests) for this new arg in https://github.com/pytorch/pytorch/blob/master/test/test_transformers.py

**Comment by PaulOppelt** [comment]
@erichan1 @mettij  On a site note: it would also be super nice, if the attention matrix could be hooked such that one also obtains the gradient of the Loss w.r.t the attention-matrix. I guess if ```need_weights```is set to true, one would still not be able to actually hook the gradient. I now went the long way around by creating a custom layer that inherits from ```TransformerEncoderLayer``` and changed ```_sa_block```. However this is not very clean

**Comment by meteorlin** [comment]
I found that it will take the first component when calculating self-attention. Does it have any specific meaning? According to my understanding, shouldn't it be to keep all the components?
In addition, assuming that this operation has practical significance, when `batch_first` is `True` and `False`, the meanings of the components obtained by `[0]` should be different (indicating `batch` and `seq` respectively), so is there any ambiguity?
https://github.com/pytorch/pytorch/blob/a6d72f44a4e8b6e9d2e878f30fd8b1d3e1197f0e/torch/nn/modules/transformer.py#L544

**Comment by kklemon** [comment]
@erichan1 
> 4. Make sure to disable fast path when need_weights==True by adding a condition in why_not_sparsity_fast_path. PyTorch transformer has a fast inference path. You could actually make need_weights work for the fast path by hooking up a couple of things, but let's skip it for now. Just default to slow path when we have this arg.

The condition should probably also include whether any hooks are attached to submodules that are affected by fast path execution. The fact that hooks seem to be ignored in this case is not really documented and cost me half a night of debugging :sweat_smile: 

<details>
<summary>See exemplary code below.</summary>

```python
import torch
import torch.nn as nn

transformer = nn.TransformerEncoder(
    nn.TransformerEncoderLayer(d_model=512, nhead=8, batch_first=True),
    num_layers=1
).cuda()


def hook(module, module_in, module_out):
    print('Hook was called')
    return module_out


transformer.layers[0].self_attn.register_forward_hook(hook)

# Training mode. Fast path is not used; hook is called
x = torch.randn(4, 32, 512).cuda()

transformer(x)

# Evaluation / inference mode. Fast path is used (sometimes?); hook is not called
transformer.eval()

with torch.no_grad():
    transformer(x)
```

</details>

**Comment by Saigum** [comment]
One thing is, the default signature of all the classess forwards which allow for returning the attention matrix is a tuple, of (ForwardOutput,AttnMatrix:Union[None,Tensor]), and for one to change it without it having variable return type(i feel like variable type output would just be a bad idea in the future), this return type would have to be enforced for all the TransformerEncoder/TransformerDecoder (with it being a list of objects for the TransformerEncoder and TransformerDecoder) and TransformerEncoderLayer,TransformerDecoderLayer and the TransformerClass itself.
